Source/Tests/LinuxSyscalls/Syscalls.h

```diff
@@ -206,6 +218,97 @@ class SyscallHandler : public FEXCore::HLE::SyscallHandler {
   #endif
 
   std::unique_ptr<FEX::HLE::MemAllocator> Alloc32Handler{};
+  
+  ///// VMA (Virtual Memory Area) tracking /////
+
+  FEXCore::HLE::AOTIRCacheEntryLookupResult LookupAOTIRCacheEntry(uint64_t GuestAddr) final override;
+
+
+  struct MRID {
+    uint64_t dev; /* kernel only uses lower 32-bits of dev_t */
+    uint64_t id;
+
+    bool operator<(const MRID& other) const {
+      return std::tie(dev, id) < std::tie(other.dev, other.id);
+    }
+  };
+
+  // We use bits 63:32 for our own custom dev tracking
+  static constexpr const uint64_t MRID_PRIVATE = 0x1'0000'0000ULL;
+  static constexpr const uint64_t MRID_ANON = 0x2'0000'0000ULL;
+  static constexpr const uint64_t MRID_SHM =  0x3'0000'0000ULL;
+
+  struct VMAEntry;
+
+  struct MappedResource {
+    typedef std::map<MRID, MappedResource> ContainerType;
+
+    FEXCore::IR::AOTIRCacheEntry *AOTIRCacheEntry;
+    VMAEntry *FirstVMA;
+    uint64_t Length; // 0 if not fixed size
+    ContainerType::iterator Iterator;
+  };
+
+  struct VMAProt {
+    bool Writable: 1;
+    bool operator==(const VMAProt& other) const {
+      return other.Writable == Writable;
+    }
+  };
+
+  struct VMAOptions {
+    VMAOptions(int Writable, const int Shared) : Prot {Writable != 0}, Flags {Shared != 0} { }
+
+    VMAProt Prot;
+
+    const struct {
+      bool Shared: 1;
+    } Flags;
+  };
+
+  struct VMAEntry {
+    MappedResource* Resource;
+    VMAEntry* ResourcePrevVMA;
+    VMAEntry* ResourceNextVMA;
+
+    uint64_t Base;
+    uint64_t Offset;
+    uint64_t Length;
+
+    VMAOptions Options;
+  };
+
+  struct VMATracking {
+    using VMAEntry = SyscallHandler::VMAEntry;
```

Comment:
The compiler needed it for the out of class implementations of `bool ListRemove(VMAEntry *Mapping);` & friends. Otherwise it needed more namespace/type stuff.

`bool SyscallHandler::VMATracking::ListRemove(VMAEntry *VMA) {` -> 
`bool SyscallHandler::VMATracking::ListRemove(SyscallHandler::VMAEntry *VMA) {` 

Funny enough, it still needs the full name in return types.

Overall, though, tbh, I'd rather see us much reduce the namespace and nested structs myself. I've found it makes the code quite a bit harder to navigate, and also /much/ harder to debug (as you need to keep using FQNs which is uh a MAJOR PAIN)

Reply:
```cpp
bool SyscallHandler::VMATracking::ListRemove(VMAEntry *VMA)
```

Are you sure the left side triggers a compiler error in this specific example? It should compile fine as by my understanding, even without the redundant type alias.

> Funny enough, it still needs the full name in return types.

C++11's trailing return type syntax resolves this issue, since it brings the names of the implemented classes in scope:

```cpp
auto SyscallHandler::VMATracking::Something() -> VMAEntry* { ... }
```

Reply:
> Are you sure the left side triggers a compiler error in this specific example? It should compile fine as by my understanding, even without the redundant type alias.

Yup, that's why I added it, because it errored out.

> C++11's trailing return type syntax resolves this issue, since it brings the names of the implemented classes in scope:

Figured it might as it knows the scope better, though ugh, I don't like how inconsistent that looks wrt everything else

Reply:
Fair enough, I guess having that convenience type alias isn't too bad then.